### ISSUE TITLE
Restructuring of towered extension fields Fp4 and Fp6

### ISSUE THREAD

**Issue by UlrichHaboeck75** [HorizenOfficial/ginger-lib#37]
As Coda we take the embedding field of an MTN4 curve as the quadratic extension 

   Fp4 = Fp2[Y] / (Y^2 - X) 

of Fp2 = Fp[X]/(X^2 - alpha), where alpha is the ``const Fp2Parameters::NONRESIDUE``, and similarly the embedding field of an MNT6 curve as 

   Fp6 = Fp2[Y]/(Y^2-X)

of Fp3=Fp[X]/(X^3 - alpha), with \alpha as ``const Fp3Parameters::NONRESIDUE``.

However, the implementations of the two towered extensions differs in style, and both 'open up' the arithmetics of their subfields Fp2, Fp3, respectively, to define multiplications by $X$ therein.
To illustrate the different styles of implementation:
- for Fp4, multiplication by Y^2 = X is done as 
```
    pub fn mul_by_nonresidue(value: &Fp2<P::Fp2Params>) -> Fp2<P::Fp2Params> {
        let new_c0 = P::Fp2Params::mul_fp_by_nonresidue(&value.c1);
        let new_c1 = value.c0;
        Fp2::new(new_c0, new_c1)
    }
```
not even referring to ``Fp2Parameters::NONRESIDUE``, whereas
- for Fp6, multiplication by $X$ is implemented as
```
    pub fn mul_by_nonresidue(value: &Fp3<P::Fp3Params>) -> Fp3<P::Fp3Params> {
        let mut res = *value;
        res.c0 = value.c2;
        res.c1 = value.c0;
        res.c2 = value.c1;
        res.c0
            .mul_assign(&<P::Fp3Params as Fp3Parameters>::NONRESIDUE);
        res
    }
```
which use ``Fp3Parameters::NONRESIDUE``, and the generic ``trait Fp6Parameters`` function 
```
   fn mul_fp3_by_nonresidue(fe: &Fp3<Self::Fp3Params>) -> Fp3<Self::Fp3Params> {
        Self::NONRESIDUE * fe
    }
``` 
is not used at all.

I propose a restructuring of the towered extension fields, choosing a generic approach which defines quadratic/cubic extensions F2/F3 over any field F (not necessarily prime), and then define F6 with "sextic" non-residue ``Fp6Params::NONRESIDUE`` as quadratic extension F2(F3), using the (quadratic) non-residue  

   (c_0,c_1)=(0,1) from F3^2
 
(which corresponds to X), where F3 is the cubic extension of Fp using the sextic non-residue as (cubic) non-residue. Likewise, one defines F4 with "quartic" non-residue ``Fp4Params::NONRESIDUE`` as quadratic extension F2(F2), using the quadratic nonresidue 

   (0,1) in F2^2,

where F2 is the quadratic extension of Fp using the quartic non-residue as quadratic one.

Such generic implementation is not just for readability - it also makes further towering (if ever needed lateron) more transparent and easy.

**Comment by UlrichHaboeck75** [comment]
The audit of the corresponding field gadgets (as part of  [this PR](https://github.com/ZencashOfficial/ginger-lib/pull/49)) shows the same structural issue: 
The FieldGadget implementations for the quadratic extensions (Fp2Gadget, Fp6Gadget as '2over3', Fp12Gadget as '2over6') as well as for the cubic extensions (Fp3Gadget, Fp6Gadget as '3 over 2') are almost identical, resulting in needless repetition of code.

I strongly recommend to implement the above described generic approach also for the gadgets:  a generic F2Gadget and F3Gadget (for arbitrary base field gadgets instead of FpGadgets only), and then defining the used Fp2Gadgets as F2Gadget over Fp, etc.

**Comment by 95DDB** [comment]
Fixed with #74, in which we imported QuadExtField/QuadExtFieldGadget and CubicExtField/CubicExtFieldGadgets structs from upstream
